### DISCUSSION TITLE
Add mapping from bazel labels to files in the runfiles manifest

### DISCUSSION THREAD

**Discussion by redsun82** [bazelbuild/bazel#20117] [Ideas]
Recently I wanted to move the body of a very complex `genrule` to a bash script wrapped in an `sh_binary` used by that `genrule`, so that I could have a simple `BUILD`. What I clashed with is how the runfiles library actually works with respect to genrules.

In genrule, you use `$(execpath @repo//a/bazel/label)`. The equivalent within the bash script using runfiles is `$(rlocation some/obscure/path/to/what/the/label/stands/for)`. This poses several challenges:

* as a rule writer, I have no idea what exact file a label refers to. There's aliases and rule outputs to unwrap, and I need to stop and research that
* the label may (quite often) refer to different files on different platforms: in that case writing a cross-platform script is a real pain

Wouldn't it be cool if the mapping from bazel labels to data files was added to the runfiles manifest, so that one could write in the bash script (or whatever else using a runfiles library) `$(rlocation @repo//a/bazel/label)` just like in a genrule? There would be the case of labels unwrapping to multiple or no files, but that should be treated as an error just like in genrules. Providing `rlocations` in bash might be delicate because of separation of items in case spaces are in the files, but I'm sure there would be solutions to that.

**Comment by fmeum** [comment]
Have you seen `$(rlocationpath(s) ...)` (https://bazel.build/reference/be/make-variables#predefined_label_variables)? While not quite as convenient, it allows you to avoid hard-coding the file paths into your code.

**Comment by armandomontanez** [comment]
This seems like it would make it easier for `rules_*` repos to manage runfiles, and could also make the API for accessing them a bit easier to hold. `rules_python`, for example, requires `r.Rlocation('repo_name/path/to/file.txt', r.CurrentRepository())`, which is neither a label, nor an actual runfiles path. As far as I can tell, the lack of correlation between labels and files means that it's not currently possible to generically implement an interface like `r.Rlocation('@repo_name//path/to:some_rule')`, and imposes limitations that make it difficult to design a user-friendly runfiles loader.

**Reply by fmeum** [reply]
The explicit `r.CurrentRepository()` argument in that call shouldn't be needed, but your point still stands.

**Reply by armandomontanez** [reply]
Perhaps I'm misunderstanding it, but I think `r.CurrentRepository()` is required for any library code that is middleware that can be consumed by a downstream project. So if you're just working on a "leaf" Bazel project, `r.CurrentRepository()` isn't necessary.

Either way, I've definitely lost hours tripping over proper usage of `rules_python`'s `Rlocation()`, and I think `rules_python` isn't particularly at fault for that since this problem is very difficult to solve right now.

**Reply by fmeum** [reply]
No, `r.Rlocation` automatically picks up the `r.CurrentRepository()` of the caller. So unless you accept a runfiles path from a (transitive) caller, you don't need to pass it in explicitly, regardless of whether you are a leaf project or a library.

**Comment by fmeum** [comment]
One approach that I tried in the past is code generation: https://github.com/fmeum/rules_runfiles?tab=readme-ov-file#step-2-accessing-the-runfiles-constants

**Reply by armandomontanez** [reply]
Thanks for the breadcrumb! I landed on a similar approach in https://pwbug.dev/388526733 when investigating this for Python.

**Comment by lorenz** [comment]
Is there anything stopping us from just having the Bazel label as an extra key in the Manifest files? That is a backwards-compatible change IMO and should be rather trivial.

**Reply by lorenz** [reply]
If there are no fundamental reasons why this shouldn't/can't be done I may have a stab at this in a few weeks.

**Reply by fmeum** [reply]
There are a number of details that I could see making this difficult:

1. Many labels do not correspond to a single files.
2. If you want aliases to be transparent, you can't use the owner label of a file as the key. Instead, you would have to somehow track all targets in the transitive closure of a target - this information isn't stored on `runfiles` yet.
3. Runfiles libraries support using the directory or the manifest, so if this information is in the manifest only, they would need to be modified.

But that doesn't mean that this can't be done and it's possible that someone needs to start prototyping this to get an accurate picture of the challenges and possibilities.

**Reply by lorenz** [reply]
Labels not corresponding to a single file are obviously much more involved, I'd skip them for now. The rest are fair points, these do indeed have significant complexity attached. The other solution would be if rules were smarter, but that only works for certain languages (for example Go has directory-scoped packages, so rules_go could just generate a file from all data dependencies and then you just have magic vars available). C/C++ could probably have these vars injected into the compiler command line. But this is obviously a per-ruleset solution.

**Reply by fmeum** [reply]
https://github.com/fmeum/rules_runfiles was more of a demo than a fully fleshed out ruleset, but I kind of liked the ergonomics, especially the compile-time validation. That of course requires per-language work, but it also wouldn't require any changes to Bazel.
